### DISCUSSION TITLE
this에 대해서 설명해 주세요

### DISCUSSION THREAD

**Discussion by kamwoo** [woowacourse-study/2022-woowahan-bansanghwe#83] [JavaScript]
## 정의

this는 현재 속해있는 객체를 가르키는 참조 값을 가르키는 변수다.

### this의 값은 런타임에 결정되기 때문에 컨텍스트에 따라 다르다

변수 선언 처리가 끝나면 다음은 this value가 결정된다. this value가 결정되기 이전에 this는 전역 객체를 가리키고 있다가 함수 호출 패턴에 의해 this에 할당되는 값이 결정된다.
전역 코드의 경우, this는 전역 객체를 가리킨다.

<img width="400" src="https://user-images.githubusercontent.com/61308364/189480298-3114b606-b306-4e27-9d64-24f104e6e62e.png" alt="this-image" />

this 프로퍼티에는 this 값이 할당된다. this에 할당되는 값은 함수 호출 패턴에 의해 결정된다.

⇒ 동일한 함수라도 다른 객체에서 호출할 때 this가 달라진다.

⇒ 메소드가 어디에서 정의됐는지와 상관없이 this가 무엇인가에 따라 자유롭게 결정된다.

### Binding 규칙

자바스크립트의 경우 함수 호출 방식에 의해 this에 바인딩할 어떤 객체가 동적으로 결정된다. 다시 말해, 함수를 선언할 때 this에 바인딩할 객체가 정적으로 결정되는 것이 아니고, 함수를 호출할 때 함수가 어떻게 호출되었는지에 따라 this에 바인딩할 객체가 동적으로 결정된다.

- 기본 바인딩
    
    기본적으로 `this`는 전역객체(Global object)에 바인딩된다. 전역함수는 물론이고 심지어 내부함수의 경우도 `this`는 외부함수가 아닌 전역객체에 바인딩된다.
    
    생성자 함수와 메소드를 제외한 메소드 내부 함수, 콜백 함수 등 모든 함수 내부의 this는 전역객체를 가리킨다. 브라우저 환경에서 비엄격 모드일 경우 this는 window, 엄격 모드에서는 undefined 이다. 
    
    내부함수는 일반 함수, 메소드, 콜백함수 어디에서 선언되었든 관게없이 this는 전역객체를 바인딩한다.
     더글라스 크락포드는 “이것은 설계 단계의 결함으로 메소드가 내부함수를 사용하여 자신의 작업을 돕게 할 수 없다는 것을 의미한다” 라고 말한다.
    
- 명시적 바인딩
    
    this를 특정 객체에 명시적으로 바인딩하는 방법도 제공된다.
    
    - call
        
        `func.call(thisArg[, arg1[, arg2[, ...]]])`
        
        `call()`메소드는 주어진 `this`값 및 각각 전달된 인수와 함께 함수를 호출한다.
        
    - apply
        
        `func.apply(thisArg, [argsArray])`
        
        `apply`
         를 사용해, 새로운 객체마다 메소드를 재작성할 필요없이 한 번만 작성해 다른 객체에 상속시킬 수 있다.
        
    - bind
        
        `func.bind(thisArg[, arg1[, arg2[, ...]]])`
        
        `bind()`함수는 새로운 바인딩한 함수를 만든다. 바인딩한 함수를 호출하면 일반적으로 래핑된 함수가 호출 된다.
        
    
- 암시적 바인딩
    
    함수가 객체의 프로퍼티 값이면 메소드로서 호출된다. 이때 메소드 내부의 `this`는 해당 메소드를 소유한 객체, 즉 해당 메소드를 호출한 객체에 바인딩된다.
    
- new 바인딩
    
    일반함수와 생성자 함수의 차이점은 호출 시 this 바인딩 방식이 다르다.
    
    일반 함수를 호출하면 this는 전역객체에 바인딩되지만 new 연산자와 함께 생성자 함수를 호출하면 this는 생성자 함수가 암묵적으로 생성한 빈 객체에 바인딩된다.
    

### 화살표 함수에서 this

화살표 함수 특징
- this나 super에 대한 바인딩이 없다.
-  call, apply, bind 메소드를 사용해도 아무런 효과가 없다
- 생성자로 사용될 수 없다.
- 함수 내부에서 yield를 사용할 수 없다.
- this, arguments가 없다.

화살표 함수는 자신의 this가 없고 함수를 둘러싸는 lexical 범위의 this가 사용된다.

⇒ 상위 스코프에 this를 찾는다.

화살표 함수는 호출되면서 생긴 실행 컨텍스트에 thisBinding을 저장하지 않는다. 때문에 실행 컨텍스트가 참조하는 this를 찾는다.

```javascript
const obj = {
  greeting (){
    return this  // obj
  }
}

const obj = {
  greeting: () => this // 전역객체 this -> obj의 this -> window
}
```

화살표 함수는 메소드 안에서 같은 this를 바라보게, 내부 함수로서 사용하는 경우에 사용

**Comment by greenblues1190** [comment]
this 바인딩의 우선순위에 대해 설명해주세요.

**Reply by kamwoo** [reply]
1. new 바인딩 
2. 명시적 바인딩
3. 암시적 바인딩
4. 기본 바인딩

new 바인딩은 새로운 객체를 만들어서 this 바인딩을 하기 때문에 가장 순위가 높다. 
`bind()`사용해서 특정 객체로 바인딩 되어 있는 함수라도, new를 사용하면 새로운 객체로 바인딩 시킨다.

**Comment by greenblues1190** [comment]
일반 함수와 화살표 함수의 차이에 대해 설명해주세요.

**Reply by kamwoo** [reply]
1. this 바인딩을 하지 않는다.
2. this 바인딩을 하지 않기 때문에 new 생성자로 사용할 수 없다.
3. this, arguments가 없다.
4. call, apply, bind를 사용해도 아무런 효과가 없다.

**Comment by greenblues1190** [comment]
렉시컬이란 어떤 것을 의미하나요?

**Reply by kamwoo** [reply]
렉시컬은 유효범위를 의미하고, 주로 식별자의 유효범위를 말할 수 있다.
코드를 실행하면서 실행 컨텍스트를 생성하는 시점에 Lexical Enviroment를 참조하는 값을 저장하게된다.
현재 스코프 내에 찾고있는 식별자가 없을 경우 참조하는 상위 스코프에서 식별자를 찾는다.

**Comment by greenblues1190** [comment]
실행 컨텍스트란 무엇인가요?

**Reply by kamwoo** [reply]
작성한 코드가 실행되기 위해서 필요한 환경을 의미한다.
실행에 필요한 정보로 변수, 함수 선언, 스코프, this value가 있다.

**Comment by greenblues1190** [comment]
실행 컨텍스트에 저장되는 데이터에는 무엇이 있나요?

**Reply by kamwoo** [reply]
- 지역 스코프내의 함수, 변수 정보를 저장하는 객체의 참조값
- 전역에서 선언된 함수, 변수를 저장하는 객체의 참조값
- 상위 Lexical Enviroment에 대한 참조값
- this value
등이 저장된다.

**Comment by greenblues1190** [comment]
클로져에 대해 설명해주세요.

**Comment by greenblues1190** [comment]
레퍼런스 카운터를 0으로 만드는 방법이 있을까요?

**Reply by kamwoo** [reply]
변수에 null 값을 할당하면 GC가 메모리에서 해제한다.

**Comment by greenblues1190** [comment]
`call`과 `apply` 메소드의 차이점에 대해 설명해주세요.

**Comment by greenblues1190** [comment]
직접 `bind` 함수를 구현하고자 합니다. 어떤 방법으로 구현할 수 있을까요?

**Comment by greenblues1190** [comment]
명시적 바인딩을 활용해본 경험이 있나요?

**Comment by greenblues1190** [comment]
말씀하신 이벤트 핸들러 바인딩은 화살표 객체로도 해결할 수 있을 것 같습니다. 명시적 바인딩을 활용한 이유가 있나요?

**Comment by greenblues1190** [comment]
자바스크립트의 클래스에 대해 설명해주세요.
